roofit/roofitcore/src/RunContextTracker.cxx

```diff
@@ -0,0 +1,71 @@
+// Author: Jonas Rembser, CERN  Feb 2021
+
+/*****************************************************************************
+ * RooFit
+ * Authors:                                                                  *
+ *   WV, Wouter Verkerke, UC Santa Barbara, verkerke@slac.stanford.edu       *
+ *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 *
+ *                                                                           *
+ * Copyright (c) 2000-2020, Regents of the University of California          *
+ *                          and Stanford University. All rights reserved.    *
+ *                                                                           *
+ * Redistribution and use in source and binary forms,                        *
+ * with or without modification, are permitted according to the terms        *
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+#include "RunContextTracker.h"
+#include "RooMsgService.h"
+
+#include "ROOT/RMakeUnique.hxx"
+
+RunContextTracker::RunContextTracker(RooBatchCompute::RunContext const &runContext)
+{
+   for (auto const &item : runContext.ownedMemory) {
+      addTracker(item.first);
+   }
+}
+
+void RunContextTracker::resetTrackers()
+{
+   for (auto const &item : trackers_) {
+      auto &tracker = *item.second;
+      tracker.hasChanged(true);
+   }
+}
+
+void RunContextTracker::cleanRunContext(RooAbsArg const &caller, RooBatchCompute::RunContext &runContext)
+{
+   for (auto const &item : trackers_) {
+      auto const *arg = item.first;
+      auto &tracker = *item.second;
+      if (tracker.hasChanged(false)) {
+         auto found = runContext.spans.find(arg);
+         if (found != runContext.spans.end()) {
+            runContext.spans.erase(found);
+         }
+      } else {
+         if (runContext.spans.find(arg) != runContext.spans.end()) {
+            oocxcoutD(&caller, FastEvaluations)
+               << "Value of " << arg->GetName() << " kept in RunContext by RunContextTracker::cleanRunContext"
+               << std::endl;
+         }
```

Comment:
Note that the map lookup will always run, irrespective of whether debug messages are on. I suggest wrapping this in
```c++
  if (RooMsgService::getInstance().msgLvl() == RooFit::DEBUG && <check message topic>) {
```

Reply:
Done, wrapped in
```C++
if (oodologD(&caller, FastEvaluations)) {    }
```